tutorials/llm/llama/domain-adaptive-pretraining/code/tokenization_helper.py

```diff
@@ -0,0 +1,400 @@
+        add_weight = torch.zeros(add_cnt, N)
+        word_embedding = torch.cat((word_embedding, add_weight), 0)
+    else:
+        old_ebd = torch.load(old_ebd_path)
+        _, N = old_ebd['word_embeddings'].shape
+        add_weight = torch.zeros(add_cnt, N)
+        old_ebd['word_embeddings'] = torch.cat((old_ebd['word_embeddings'], add_weight), 0)
+
+    if split > 1:
+        _, M = output_layer.shape
+        add_out = torch.zeros(add_cnt, M)
+        output_layer = torch.cat((output_layer, add_out), 0)
+    else:
+        _, M = old_ebd['output_layer'].shape
+        add_out = torch.zeros(add_cnt, M)
+        old_ebd['output_layer'] = torch.cat((old_ebd['output_layer'], add_out), 0)
+
+    sp = spm.SentencePieceProcessor()
+    sp.load(original_tokenizer_path)
+
+    for r in record:
+        token = r[0]
+        idx = r[1]
+        ids = sp.encode_as_ids(token)
+        if split > 1:
+            word_embedding[idx] = torch.mean(word_embedding[ids], dim=0)
+            output_layer[idx] = torch.mean(output_layer[ids], dim=0)
+        else:
+            old_ebd['word_embeddings'][idx] = torch.mean(old_ebd['word_embeddings'][ids], dim=0)
+            old_ebd['output_layer'][idx] = torch.mean(old_ebd['output_layer'][ids], dim=0)
+
+    if split > 1:
+        KK, K = word_embedding.shape
+        print("embedding size: ", K)
+        print("vocab size: ", KK)
+        print("#split: ", split)
+        T = K // split
+        R = KK // split
+        prefix = new_ebd_path + "/embedding_"
+        for i in range(split):
+            start = i * T
+            end = (i + 1) * T
+            st = i * R
+            ed = (i + 1) * R
+            save_name = prefix + f"{i}" + ".pt"
+            temp = {}
+            temp['word_embeddings'] = word_embedding[:, start:end]
+            temp['output_layer'] = output_layer[st:ed, :]
+            # print("split word_embedding shape: ", temp['word_embeddings'].shape)
+            # print("split output_layer shape: ", temp['output_layer'].shape)
+            torch.save(temp, save_name)
+    else:
+        torch.save(old_ebd, new_ebd_path + str(len(m.pieces)) + ".pt")
+
+    print("Completed saving new embeddings")
+
+
+def analyze_token_usage(data_root, tokenizer_path, batchsize, keys, save_path):
+    """
+    Function to analyze domain tokens using frequency analysis
+    Args:
+            data_root (str): Path to general/domain specific data to identify tokens
+            tokenizer_path (str): Path to original tokenizer (llama 2 tokenizer downlaoded from hf)
+            batchsize (int): batch size used for the text_iterator that generates of batches of text.
+            keys (list): Keys/metadata to extract from jsonl files
+            save_path (str): path to save token usage frequency analysis results
+
+    Returns:
+           None, saves frequency analysis results to the provided path
+
+    """
+    # extra_id = 255000
+    extra_id = 32000  ## for mixtral tokenizer
```

Comment:
Is this "for mixtral tokenizer" comment necessary here?

Reply:
No, its not strictly needed. I have removed it.